src/p11_front.c

```diff
@@ -367,18 +369,34 @@ int PKCS11_set_ui_method(PKCS11_CTX *pctx, UI_METHOD *ui_method, void *ui_user_d
 	return pkcs11_set_ui_method(ctx, ui_method, ui_user_data);
 }
 
-/* External interface to the deprecated features */
-
-int PKCS11_generate_key(PKCS11_TOKEN *token,
-		int algorithm, unsigned int bits,
-		char *label, unsigned char *id, size_t id_len)
+int PKCS11_generate_key(PKCS11_TOKEN *token, PKCS11_KGEN_ATTRS *kg)
 {
 	PKCS11_SLOT_private *slot = PRIVSLOT(token->slot);
 	if (check_slot_fork(slot) < 0)
 		return -1;
-	return pkcs11_generate_key(slot, algorithm, bits, label, id, id_len);
+	unsigned char out[128] = {0};
+	size_t key_id_len = 0;
+	if (kg && kg->key_id) {
+		key_id_len = strlen(kg->key_id);
+		if (key_id_len > 127) {
```

Comment:
While this is not a security problem, should hex to bin compress the output by factor 1/2?

Reply:
Do you mean output length? For even lengths yes, for odd no.
0x10a (3 bytes)  ->  0000 0001 0000 1010 (2 bytes)

Reply:
```suggestion
		if (key_id_len == 127) {
```
> RETURN VALUE
>        The strnlen() function returns strlen(s), if that is less than maxlen, or maxlen if there is no null termi‐
>        nating ('\0') among the first maxlen characters pointed to by s.

I wonder why something greater than 127 shall be returned in any case.